engine/runtime/src/memories/frag.rs

```diff
@@ -0,0 +1,624 @@
+        unsafe {
+            let ptr: *mut T = self.ptr.as_mut();
+            *ptr = T::default();
+        }
+    }
+}
+
+impl<T: Default + Clone> Drop for Frag<T> {
+    fn drop(&mut self) {
+        self.zeroize();
+        Frag::dealloc(self).expect("Error while deallocating fragment memory");
+    }
+}
+
+unsafe impl<T: Default + Clone + Send> Send for Frag<T> {}
+unsafe impl<T: Default + Clone + Sync> Sync for Frag<T> {}
+
+/// Configuration for the fragmenting allocator
+pub struct FragConfig {
+    /// The last address of a previous allocation. This
+    /// value will be used to calculate the minimum distance to the
+    /// previous allocation.
+    pub(crate) last_address: usize,
+
+    /// The  minimum distance to a previous allocation
+    pub(crate) min_distance: usize,
+}
+
+impl FragConfig {
+    /// Creates a new [`FragConfig`]
+    pub fn new(last_address: usize, min_distance: usize) -> Self {
+        Self {
+            last_address,
+            min_distance,
+        }
+    }
+}
+
+impl<T: Default + Clone> Frag<T> {
+    /// Returns a fragmenting allocator by strategy
+    ///
+    /// # Example
+    ///
+    /// ```skip
+    /// use stronghold_engine::runtime::memories::*;
+    ///
+    /// let object  = Frag::by_strategy(FragStrategy::Default).unwrap();
+    /// ```
+
+    /// Tries to allocate two objects of the same type with a minimum distance in memory space.
+    pub fn alloc2(strategy: FragStrategy, distance: usize) -> Result<(Frag<T>, Frag<T>), MemoryError> {
+        let a = match strategy {
+            FragStrategy::Direct => DirectAlloc::alloc(None),
+            FragStrategy::Map => MemoryMapAlloc::alloc(None),
+            FragStrategy::Hybrid => DirectAlloc::alloc(None),
+        }?;
+
+        let config = Some(FragConfig::new(a.ptr.as_ptr() as usize, distance));
+        let b = match strategy {
+            FragStrategy::Direct => DirectAlloc::alloc(config),
+            FragStrategy::Map => MemoryMapAlloc::alloc(config),
+            FragStrategy::Hybrid => MemoryMapAlloc::alloc(config),
+        }?;
+
+        let actual_distance = calc_distance(a.get()?, b.get()?);
+        if actual_distance < distance {
+            error!(
+                "Distance between parts below threshold: \na: 0x{:016x} \nb: 0x{:016x} \ngiven_value: 0x{:016x}",
+                a.ptr.as_ptr() as usize,
+                b.ptr.as_ptr() as usize,
+                &a as *const _ as usize
+            );
+            return Err(MemoryError::Allocation(format!(
+                "Distance between parts below threshold: 0x{:016X}",
+                actual_distance
+            )));
+        }
+
+        info!("Mapped 2 fragments: at {:?} and {:?}", a.ptr, b.ptr);
+        Ok((a, b))
+    }
+
+    /// Tries to allocate two objects of the same type with a default minimum distance in memory space of
+    /// `FRAG_MIN_DISTANCE`.
+    pub fn alloc(strategy: FragStrategy, data1: T, data2: T) -> Result<(Frag<T>, Frag<T>), MemoryError> {
```

Comment:
The naming `alloc2` is a remnant of the past and returned two fragments, while `alloc` returned one. I would rename the functions to

- `alloc` -> `alloc_initialized`
- `alloc2` -> `alloc` or `alloc_default`